crates/utils/src/interval_map/mod.rs

```diff
@@ -0,0 +1,1006 @@
+    where
+        R: 'a,
+        F: FnOnce(&'a mut Node<T, V, Ix>) -> R,
+    {
+        op(&mut self.nodes[node.index()])
+    }
+
+    fn left_ref<F, R>(&'a self, node: NodeIndex<Ix>, op: F) -> R
+    where
+        R: 'a,
+        F: FnOnce(&'a Node<T, V, Ix>) -> R,
+    {
+        let idx = self.nodes[node.index()].left().index();
+        op(&self.nodes[idx])
+    }
+
+    fn right_ref<F, R>(&'a self, node: NodeIndex<Ix>, op: F) -> R
+    where
+        R: 'a,
+        F: FnOnce(&'a Node<T, V, Ix>) -> R,
+    {
+        let idx = self.nodes[node.index()].right().index();
+        op(&self.nodes[idx])
+    }
+
+    fn parent_ref<F, R>(&'a self, node: NodeIndex<Ix>, op: F) -> R
+    where
+        R: 'a,
+        F: FnOnce(&'a Node<T, V, Ix>) -> R,
+    {
+        let idx = self.nodes[node.index()].parent().index();
+        op(&self.nodes[idx])
+    }
+
+    fn grand_parent_ref<F, R>(&'a self, node: NodeIndex<Ix>, op: F) -> R
+    where
+        R: 'a,
+        F: FnOnce(&'a Node<T, V, Ix>) -> R,
+    {
+        let parent_idx = self.nodes[node.index()].parent().index();
+        let grand_parent_idx = self.nodes[parent_idx].parent().index();
+        op(&self.nodes[grand_parent_idx])
+    }
+
+    fn left_mut<F, R>(&'a mut self, node: NodeIndex<Ix>, op: F) -> R
+    where
+        R: 'a,
+        F: FnOnce(&'a mut Node<T, V, Ix>) -> R,
+    {
+        let idx = self.nodes[node.index()].left().index();
+        op(&mut self.nodes[idx])
+    }
+
+    fn right_mut<F, R>(&'a mut self, node: NodeIndex<Ix>, op: F) -> R
+    where
+        R: 'a,
+        F: FnOnce(&'a mut Node<T, V, Ix>) -> R,
+    {
+        let idx = self.nodes[node.index()].right().index();
+        op(&mut self.nodes[idx])
+    }
+
+    fn parent_mut<F, R>(&'a mut self, node: NodeIndex<Ix>, op: F) -> R
+    where
+        R: 'a,
+        F: FnOnce(&'a mut Node<T, V, Ix>) -> R,
+    {
+        let idx = self.nodes[node.index()].parent().index();
+        op(&mut self.nodes[idx])
+    }
+
+    fn grand_parent_mut<F, R>(&'a mut self, node: NodeIndex<Ix>, op: F) -> R
+    where
+        R: 'a,
+        F: FnOnce(&'a mut Node<T, V, Ix>) -> R,
+    {
+        let parent_idx = self.nodes[node.index()].parent().index();
+        let grand_parent_idx = self.nodes[parent_idx].parent().index();
+        op(&mut self.nodes[grand_parent_idx])
+    }
+}
+
+/// An iterator over the entries of a `IntervalMap`.
+#[derive(Debug)]
+pub struct Iter<'a, T, V, Ix> {
+    /// Reference to the map
+    map_ref: &'a IntervalMap<T, V, Ix>,
+    /// Stack for iteration
+    stack: Option<Vec<NodeIndex<Ix>>>,
+}
+
+impl<T, V, Ix> Iter<'_, T, V, Ix>
+where
+    Ix: IndexType,
+{
+    /// Initializes the stack
+    fn init_stack(&mut self) {
+        self.stack = Some(Self::left_tree(self.map_ref, self.map_ref.root));
+    }
+
+    /// Pushes x and x's left sub tree to stack.
+    fn left_tree(map_ref: &IntervalMap<T, V, Ix>, mut x: NodeIndex<Ix>) -> Vec<NodeIndex<Ix>> {
```

Comment:
Why do you only iterate the left sub tree?

Reply:
Better rename this method and modify the comment

Reply:
renamed to `left_link`, the current naming is misleading